src/frontend/src/table/insert.rs

```diff
@@ -13,79 +13,54 @@
-use datatypes::prelude::ConcreteDataType;
-use snafu::{ensure, OptionExt, ResultExt};
+use datatypes::prelude::{ConcreteDataType, VectorRef};
+use futures::future;
+use snafu::{ensure, ResultExt};
 use store_api::storage::RegionNumber;
 use table::requests::InsertRequest;
 
 use super::DistTable;
 use crate::error;
-use crate::error::{FindTableRouteSnafu, Result};
-use crate::table::scan::DatanodeInstance;
+use crate::error::{JoinTaskSnafu, RequestDatanodeSnafu, Result};
 
 impl DistTable {
-    pub async fn dist_insert(
-        &self,
-        inserts: HashMap<RegionNumber, InsertRequest>,
-    ) -> Result<Output> {
-        let table_name = &self.table_name;
-        let route = self
-            .partition_manager
-            .find_table_route(&self.table_name)
-            .await
-            .with_context(|_| FindTableRouteSnafu {
-                table_name: table_name.to_string(),
-            })?;
-
-        let mut joins = Vec::with_capacity(inserts.len());
-        for (region_id, insert) in inserts {
-            let datanode = route
-                .region_routes
-                .iter()
-                .find_map(|x| {
-                    if x.region.id == region_id as u64 {
-                        x.leader_peer.clone()
-                    } else {
-                        None
-                    }
+    pub async fn dist_insert(&self, inserts: Vec<GrpcInsertRequest>) -> Result<Output> {
+        let regions = inserts.iter().map(|x| x.region_number).collect();
+        let instances = self.find_datanode_instances(regions).await?;
+
+        let results = future::try_join_all(instances.into_iter().zip(inserts.into_iter()).map(
+            |(instance, request)| {
+                common_runtime::spawn_write(async move {
+                    instance
+                        .grpc_insert(request)
+                        .await
+                        .context(RequestDatanodeSnafu)
                 })
-                .context(error::FindDatanodeSnafu { region: region_id })?;
-
-            let client = self.datanode_clients.get_client(&datanode).await;
-            let db = Database::new(&table_name.catalog_name, &table_name.schema_name, client);
-            let instance = DatanodeInstance::new(Arc::new(self.clone()) as _, db);
-
-            let join = common_runtime::spawn_write(async move {
-                instance
-                    .grpc_insert(to_grpc_insert_request(region_id, insert)?)
-                    .await
-                    .context(error::RequestDatanodeSnafu)
-            });
-
-            joins.push(join);
-        }
+            },
+        ))
+        .await
+        .context(JoinTaskSnafu)?;
 
-        let mut success = 0;
-        for join in joins {
-            let rows = join.await.context(error::JoinTaskSnafu)?? as usize;
-            success += rows;
-        }
-        Ok(Output::AffectedRows(success))
+        let affected_rows = results.into_iter().collect::<Result<Vec<_>>>()?;
```

Comment:
This might work https://doc.rust-lang.org/stable/std/result/enum.Result.html#impl-Sum%3CResult%3CU,+E%3E%3E-for-Result%3CT,+E%3E